spur/ssh.py

```diff
@@ -80,7 +80,14 @@ def generate_run_command(self, command_args, store_pid,
             commands.append("echo $$")
 
         if cwd is not None:
-            commands.append("cd {0}".format(escape_sh(cwd)))
+            # 1. Change directory before to check if the command exists
+            # (using command || which) because it might be in this
+            # directory:
+            # $ shell.run(["./foo"], cwd="/some/dir")
+            # 2. If we exit when this cd fails, we can't capture the which_return_code.
+            # To minimize the changes, we run cd again just before exec.
+            # Send stderr to /dev/null to capture only the second cd.
+            commands.append("cd {0} 2>/dev/null".format(escape_sh(cwd)))
```

Comment:
Do you have something nicer than that:
```
cd '/some/path'; status=$?; echo $status; [ $status -eq 0 ] && { command -v 'ls' > /dev/null
2>&1 || which 'ls' > /dev/null 2>&1; echo $?; exec 'ls'; }
```
to get the status code of cd and don't run the command if it fails?